### DISCUSSION TITLE
METplus StatAnalysis Template Error

### DISCUSSION THREAD

**Discussion by lindsayrblank** [dtcenter/METplus#1328] [Configuration] [answered]
Hello again, 

I am running the metplus_4.1.0-beta4 Docker image on an HPC using Singularity. I am running StatAnalysis in my use case. The relevant code snippet is here:

`# stat_analysis job info`
`STAT_ANALYSIS_JOB_NAME = aggregate_stat`
`# if using -dump_row, put in JOBS_ARGS "-dump_row [dump_row_file]"`
`# if using -out_stat, put in JOBS_ARGS "-out_stat [out_stat_file]"`
`# METplus will fill in filename`
`STAT_ANALYSIS_JOB_ARGS = -out_line_type CTC -out_thresh '>0' -by OBS_SID,FCST_VAR,FCST_INIT_HOUR -out_stat` `[out_stat_file]`

I have set the following template: 

`STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{valid?fmt=%y%m%d}_fcstlead{lead?fmt=%H}_ctc.stat`

When I run, the output file is named as such: `modelA__fcstlead120000_ctc.stat`

Other templates work just fine i.e.
`FCST_PCP_COMBINE_OUTPUT_TEMPLATE = {init?fmt=%Y%m%d}/{init?fmt=%H}/upp.t00z.pgrb2.0p125.{valid?fmt=%Y%m%d%H}_combined`

Why is this template not working?

Thank you,
Lindsay

**Comment by j-opatz** [comment]
Hi Lindsay,

And thank you for your question. It looks like the template that's causing the issues may be set with a syntax error on the date; specifically, the `%y` should be a `%Y`. Try

`STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{valid?fmt=%Y%m%d}_fcstlead{lead?fmt=%H}_ctc.stat`

and see if that fixes the issue.

**Reply by lindsayrblank** [reply]
Hi John,

Thanks for catching that. I used `%Y` and I still get the same out file name. 

`STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{valid?fmt=%Y%m%d}_fcstlead{lead?fmt=%H}_ctc.stat`

`modelA__fcstlead120000_ctc.stat`

**Reply by malloryprow** [reply]
Hi Lindsay, Could you try 

STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{fcst_valid?fmt=%Y%m%d}_fcstlead{lead?fmt=%H}_ctc.stat

Depending on settings in your config file, "valid" might but be set by the StatAnalysis wrapper.

**Reply by lindsayrblank** [reply]
Hi Mallory,

Thanks for the suggestion, but unfortunately I still get `modelA__fcstlead120000_ctc.stat` as the output. Any other ideas?

**Reply by malloryprow** [reply]
Are you able to attach your config file somewhere here so I can look at it?

**Reply by malloryprow** [reply]
For the StatAnalysis wrapper, you can use the following for date information in the template:

special_keys = [
            'fcst_valid_hour_beg', 'fcst_valid_hour_end',
            'fcst_init_hour_beg', 'fcst_init_hour_end',
            'obs_valid_hour_beg', 'obs_valid_hour_end',
            'obs_init_hour_beg', 'obs_init_hour_end',
            'valid_hour', 'valid_hour_beg', 'valid_hour_end',
            'init_hour', 'init_hour_beg', 'init_hour_end',
            'fcst_valid', 'fcst_valid_beg', 'fcst_valid_end',
            'fcst_init', 'fcst_init_beg', 'fcst_init_end',
            'obs_valid', 'obs_valid_beg', 'obs_valid_end',
            'obs_init', 'obs_init_beg', 'obs_init_end',
            'valid', 'valid_beg', 'valid_end',
            'init', 'init_beg', 'init_end',
            'fcst_lead_hour', 'fcst_lead_min',
            'fcst_lead_sec', 'fcst_lead_totalsec',
            'obs_lead_hour', 'obs_lead_min',
            'obs_lead_sec', 'obs_lead_totalsec',
            'lead', 'lead_hour', 'lead_min', 'lead_sec', 'lead_totalsec'
        ]

**Reply by lindsayrblank** [reply]
Sure, here it is. I gave it a ".txt" suffix so GitHub would upload it. 
[PCPCombine_PointStat_StatAnalysis_ModelA_ctc.conf.txt](https://github.com/dtcenter/METplus/files/7747425/PCPCombine_PointStat_StatAnalysis_ModelA_ctc.conf.txt)

**Reply by lindsayrblank** [reply]
I just saw your reply! That is a handy list to know. However, I was using 'valid' and it still didn't work so I'm not sure what's going on.

**Reply by lindsayrblank** [reply]
Hi @malloryprow and Happy New Year! I was wondering if you've had a chance to investigate this issue further?

**Reply by malloryprow** [reply]
@lindsayrblank I haven't, I'm sorry! Have you tried any of the other options in place of valid?

**Reply by lindsayrblank** [reply]
I've used 'fcst_valid' and 'valid'. The file format I need to use requires the forecast valid date in the file name. 

'lead' works, interestingly enough, but not 'valid' or 'fcst_valid'.

**Comment by georgemccabe** [comment, marked as answer]
Hi @lindsayrblank , StatAnalysis wrapper is not like many of the other wrappers where you can freely switch between init and valid in the filename templates. If you set LOOP_BY = INIT, you can reference 'init' template tags but not 'valid' ones. The wrapper does not loop over the forecast leads and process once for each init/lead or valid/lead combination. It actually just runs once and passes in the lead list to the STATAnalysis config file. Because of this, there is no way to compute what the corresponding valid time "should" be when "looping" by INIT. Switching your filename template for the output file to reference init instead of valid should fix the issue:

STAT_ANALYSIS_OUT_STAT_TEMPLATE = {model?fmt=%s}_{init?fmt=%Y%m%d}_fcstlead{lead?fmt=%H}_ctc.stat

**Reply by lindsayrblank** [reply]
Thanks for the explanation, @georgemccabe. Unfortunately, that template you gave me does not work either. Neither does substituting in 'fcst_init' or changing the format to '%Y%m%d%H'. Any ideas?

**Reply by georgemccabe** [reply]
Try using 'fcst_init_beg' or 'fcst_init_end'

**Reply by lindsayrblank** [reply]
'fcst_init_beg' worked, thank you. I can work with that going forward.
